sgrep/tests/python/equivalence_f_string.py

```diff
@@ -0,0 +1,30 @@
+def foo1():
+  # ERROR:
+  w = "foo"
+  query = f"hello {w} hel"
+
+def foo2():
+  # ERROR:
+  ww = "bar"
+  query = f"ASD{ww}ASASD"
+
+def foo3():
+  # OK:
+  www = "bar"
+  query = f"SELECT {www}"
+
+def foo4():
+  # OK:
```

Comment:
Shouldn't this match your pattern?

Reply:
`...` is matching more than 0 characters. Something like
```
$M = "..."
$N = "..."
...
$Q = f"...{$M}...{$N}"
``` matches though!

Reply:
Hmm, I'd think constant propagation should treat these as equivalent. Maybe a bug to file?

Reply:
Honestly I think `...` should match 0 characters.

Reply:
"..." matches 0 characters.
The issue here is that f"...{$M}"..." is internally rewritten as
special_concat("...", $M, "...") 
and a string like f"foo{a}" is internally rewritten as
special_concat("foo", a)
and the issue here is that the number of arguments are different so the engine will say "no match".

We need to handle specially calls to special_concat and accept that "..." matches not only a string, or an empty string, or nothing at all.
In fact we probably want f"...{$X}..." to match any interpolated expression, not just the first one, so you also don't have to write another pattern when an f-string use 2 interpolated expressions as in f"one {a} other stuff {b}"

Reply:
@aryx: I have update the generic matcher such that  "..." could match many more arguments, thus taking care of the case  where  the number of arguments are different.
However, I use `constant_propagation_and_evaluate_literal` to determine if variable `www` is a literal.  As the moment, the constant propagation in python does not seem to keep track of `www` as a literal